test/testNodes.ts

```diff
@@ -0,0 +1,782 @@
+	interface ASTReturn extends ASTNode {}
+	interface ASTImportIdent extends ASTNode {}
+	interface ASTImport extends ASTNode {}
+	interface ASTIfExpr extends ASTNode {}
+	interface ASTElifExpr extends ASTNode {}
+	interface ASTElseExpr extends ASTNode {}
+	interface ASTIfStmt extends ASTNode {}
+	interface ASTForStmt extends ASTNode {}
+	interface ASTWhileStmt extends ASTNode {}
+	interface ASTVisibility extends ASTNode {}
+	interface ASTParams extends ASTNode {}
+	interface ASTReturnType extends ASTNode {}
+	interface ASTTemplate extends ASTNode {}
+	interface ASTFunction extends ASTNode {}
+	interface ASTOperator extends ASTNode {}
+	interface ASTClass extends ASTNode {}
+	interface ASTBlock extends ASTNode {}
+}
+
+declare module '../src/server/ast/literals'
+{
+	interface ASTInt extends ASTNode {}
+	interface ASTFloat extends ASTNode {}
+	interface ASTStringLit extends ASTNode {}
+	interface ASTCharLit extends ASTNode {}
+	interface ASTBool extends ASTNode {}
+	interface ASTNull extends ASTNode {}
+}
+
+declare module '../src/server/ast/operations'
+{
+	interface ASTFunctionCall extends ASTNode {}
+	interface ASTUnaryOp extends ASTNode {}
+	interface ASTBinaryOp extends ASTNode {}
+	interface ASTPrefixOp extends ASTUnaryOp {}
+	interface ASTPostfixOp extends ASTUnaryOp {}
+	interface ASTDeref extends ASTUnaryOp {}
+	interface ASTInvert extends ASTUnaryOp {}
+	interface ASTMul extends ASTUnaryOp {}
+	interface ASTAdd extends ASTUnaryOp {}
+	interface ASTShift extends ASTUnaryOp {}
+	interface ASTBit extends ASTUnaryOp {}
+	interface ASTRel extends ASTUnaryOp {}
+	interface ASTBetween extends ASTNode {}
+	interface ASTLogic extends ASTUnaryOp {}
+	interface ASTAssign extends ASTNode {}
+}
+
+function eqOk()
+{
+	return Ok({
+		message: '',
+		pass: true,
+	})
+}
+
+function eqFail(message: string, path: TestPath)
+{
+	let errorLocation = ''
+	if (path.latestToken)
+	{
+		const line = path.latestToken.location.start.line
+		const column = path.latestToken.location.start.character
+		errorLocation = `At line: ${line + 1}, column: ${column}\n`
+	}
+
+	return Err({
+		message: `${errorLocation}${message} in ${path.path}`,
+		pass: false,
+	})
+}
+
+type NextT = {toString: () => string}
+function nextPath<T extends NextT>(path: TestPath, expected: T): TestPath
+{
+	const newPath = {...path}
+	newPath.path = `${path.path} -> ${expected.toString()}`
+	return newPath
+}
+
+function matchTypeAndToken(received: ASTNode, expected: ASTNode, path: TestPath): EqualPass
+{
+	path.latestToken = received.token
+
+	if (received.type !== expected.type)
+	{
+		const msg = `Expected ASTType: '${ASTType[expected.type]}', received '${ASTType[received.type]}'`
+		return eqFail(msg, path)
+	}
+
+	received.token.isEqualNode(expected.token, path).unwrap()
+	return eqOk()
+}
+
+type EqualType<T> = {isEqualNode: (expected: T, path: TestPath) => EqualPass} & NextT
+function undefinedEqual<T extends EqualType<T>>(
+	received: T | undefined,
```

Comment:
Can this not be `received?: T`? Allowing this to all be smushed back onto one line..

If not, then `path: TestPath,` should be on its own line too

Reply:
`received?: T` can be only used when argument is optional and optional arguments have to be after required arguments. Using `T | undefined`, the function APIs of `undefinedEqual`,  `undefinedListEqual`, and `listEqual` are consistent